lib/spack/spack/build_environment.py

```diff
@@ -165,7 +166,14 @@ def set_compiler_environment_variables(pkg, env):
     env_flags = {}
     build_system_flags = {}
     for flag in spack.spec.FlagMap.valid_compiler_flags():
-        injf, envf, bsf = pkg.flag_handler(flag, pkg.spec.compiler_flags[flag])
+        if isinstance(pkg.flag_handler, types.FunctionType):
```

Comment:
This looks like it preps for the following possibilities:

* A package implements `flag_handler`
* A package sets `flag_handler = ...`

Are both approaches tested? If not, should they both be tested?

Also IMO it's worth mentioning what this is addressing here - e.g. add a comment like 

```
# Address the case where the package chooses a flag handler using a statement like "flag_handler =..."
```

Reply:
There are three types of "function" like objects in python2, and two types in python3. 

In python2, there are 2 types of methods, "bound" and "unbound", and one type "function". The `method.im_func` attribute returns a function type object from either a bound or unbound method. I previously tested only converting from a bound method, although the conversion is the same. I've added a test for unbound methods.

In python3 there are not unbound methods. The testing already covers both functions and methods in python3 speak.

I don't think an additional test was necessary, because the conversion to function type works for both bound and unbound methods, so once we changed everything to function type they behave identically. However, writing the additional test took way less time than writing this comment, so I just added it anyway. More tests can't hurt.

Reply:
@becker33 sorry could you also add a comment here mentioning that this addresses the different possible ways a user can set up their `flag_handler` method? Other than that this LGTM